Postgresql changes for single and flexible

wjpigott opened the pull request:
terraform commands are creating the database workbook properly but I am receiving a syntax error after the ./deploy-workbook.sh -s 08b0af8b-8743-42cc-b219-e440a64f67a1 -g postgresql.  I think the workbook is correct now using the new extend kql commands.

tsubasaxZZZ commented:
I've updated MySQL part, so, I'd like you to revise PostgreSQL part based on it.

Please make sure the following items:

- [ ] 1. Move the Summary part under Single pane.
- [ ] 2. For Summary part, only Flexible.
- [ ] 3. Add notification about Single Server being retired.
- [ ] 4. Adjust column MySQL to ether Single and Flexible.
- [ ] 5. Make sure burstable SKU is production SKU

![image](https://github.com/Azure/reliability-workbook/assets/3822284/f6284422-0ad5-445b-b73f-b5e74b686588)

cf MySQL.:
![image](https://github.com/Azure/reliability-workbook/assets/3822284/58af9593-51d0-4842-9ef8-5b0d018d562c)

wjpigott commented:
@tsubasaxZZZ for the question on Burstable if this would be considered production, I wouldnt think so as it is triggered when you change the server to be development. 
![image](https://github.com/Azure/reliability-workbook/assets/86477713/a757ce2c-6206-4d40-9a5d-edcf92ea84e9)

wjpigott commented:
I changed my label from GeoRedundantBacup to Geo Backup and changed Zone to "AvailabiltyZone".  Some labels have spaces and others dont have spaces so not sure the pattern there. 

As far as the labels on the columns and order of them. I have reordered Zone to be in the same order.  If you want more changed please specify exactly the column and name you would like it to be. I have tried to match as many as I see that makes sense for the service. 

Added area for warning for Single server to point to  https://learn.microsoft.com/en-us/azure/postgresql/single-server/whats-happening-to-postgresql-single-server

I moved the summary to be under Single Server, and changed the query so it is only including Flexible server now. 
![image](https://github.com/Azure/reliability-workbook/assets/86477713/904f50b6-e10b-4632-9d5c-d6c5329b290d)
@tsubasaxZZZ

wjpigott commented:
@tsubasaxZZZ pushed my changes up, let me know if you see something else. For the summary area, I am not sure there are other properties to add, if you see something please let me know.

tsubasaxZZZ commented:
@wjpigott Thank you for your update, and I reviewed your update again.

- PostgreSQL part is not considered replica. And it need to fix small update. I attach updated MySQL tab, so, please refer to this by importing it.
  - [FTA - Reliability Workbook - v2.9.1 - 0821-databases (1).txt](https://github.com/Azure/reliability-workbook/files/12672316/FTA.-.Reliability.Workbook.-.v2.9.1.-.0821-databases.1.txt)
  - ![image](https://github.com/Azure/reliability-workbook/assets/3822284/d63e25ca-d24c-42c7-8070-6d3f62ef8962)
- The following files should be remove to commit.
![image](https://github.com/Azure/reliability-workbook/assets/3822284/e29e1162-5785-4bbf-9012-7466505e27f0)

wjpigott commented:
![image](https://github.com/Azure/reliability-workbook/assets/86477713/e2401616-73a3-4614-b652-7ace71c2c6ad)
Added replication role and section in the summary. I am not sure what to summarize on if it is number of primary or ?

wjpigott commented:
These changes were updated in GIT as well so should be good for you to look at. @tsubasaxZZZ

tsubasaxZZZ commented on build/template_kql/database/database_postgresql_resources_details_summary.kql:
Rename file name from `database_postgresql_resources_details_summary.kql` to `database_postgresqlflexible_resources_details_summary.kql`

wjpigott replied:
implemented this change

tsubasaxZZZ commented on build/templates/databases.tpl.json:
Please check and change MySQL UI to maintain consistency.

wjpigott replied:
Mysql example:
![image](https://github.com/Azure/reliability-workbook/assets/86477713/cf3744c4-0740-40b6-96f3-498437f60da0)

Postgresql example:
![image](https://github.com/Azure/reliability-workbook/assets/86477713/af61e3f2-da9a-4699-83be-299b67a4a6ad)

I think the columns are 1-1 here between. Does this look good @tsubasaxZZZ

wjpigott replied:
So the remaining item to correct is the replica count. I'll talk to some resources to see if they are aware of a way to get this between primary and asyncreplica counts. @tsubasaxZZZ

wjpigott replied:
All other changes have been updated or checked in.

wjpigott replied:
We would need to do some kind of subtraction of the 2 counts to get the No Replica numbers. This works just shows 2 extra columns, maybe we can hide them somehow.

![image](https://github.com/Azure/reliability-workbook/assets/86477713/838e833e-c763-4bcf-ba1c-2cc76aabc3aa)

wjpigott replied:
This json will hide the 2 columns so the only issue would be the order of Read Replica count being at the end.       
  {
          "columnMatch": "Primary",
          "formatter": 5
        },
        {
          "columnMatch": "AsyncReplica",
          "formatter": 5
        }

wjpigott replied:
![image](https://github.com/Azure/reliability-workbook/assets/86477713/9a975f24-c33c-4727-b383-77bac1b5e7ce)
This should be accurate.

tsubasaxZZZ replied:
Please check the following discussion:
https://github.com/Azure/reliability-workbook/pull/135#discussion_r1333882508

tsubasaxZZZ reviewed (changes requested):
Please check my comment.

wjpigott commented on build/template_kql/database/database_postgresql_resources_details_summary.kql:
Replication Role for MySQL is     "replicationRole": "None",
Replication Role for Postgresql is not None,  it is either  "replicationRole": "AsyncReplica",     "replicationRole": "Primary",

So we are not able to summarize on "None" as this doesnt exist.   @tsubasaxZZZ

wjpigott replied:
This is what I have now that is subtracting the Primary - asyncreplica to get the NoReplica numbers

| summarize ResourceTotal=count(Name), NotRunning=dcountif(Name, State !in~ ("ready")), NotRecommendedSKU=dcountif(Name, SkuName == "burstable"), NoAVZone=dcountif(Name, AvailabilityZone == "not configured"), NoHA=dcountif(Name, AutomaticFailover == "not configured"), NoGeoBackup=dcountif(Name, GeoBackup == "disabled"), Primary=dcountif(Name, ReplicationRole == "primary"), AsyncReplica=dcountif(Name, ReplicationRole == "asyncreplica")
| extend NoReplica = Primary - AsyncReplica

tsubasaxZZZ replied:
Please check the following discussion:
https://github.com/Azure/reliability-workbook/pull/135#discussion_r1333882508

wjpigott commented on build/template_kql/common/extend_resource.kql:
```diff
@@ -149,6 +155,8 @@ resources
     (type == 'microsoft.documentdb/databaseaccounts'), iif(properties.locations[0].isZoneRedundant == "false", 'Not Configured', 'Configured'), 
     (type == 'microsoft.dbformysql/servers'), "Not Applicable",
     (type == 'microsoft.dbformysql/flexibleservers'), iif(properties.haEnabled == "Enabled", 'Configured', 'Not Configured'),
+    (type == 'microsoft.dbforpostgresql/servers'), "Not Applicable",
+    (type == 'microsoft.dbforpostgresql/flexibleservers'), coalesce(tostring(array_length(parse_json(zones))), 'Not Configured'),
```
implemented this change

wjpigott commented on build/template_kql/common/extend_resource.kql:
```diff
@@ -170,6 +178,7 @@ resources
     (type == "microsoft.recoveryservices/vaults"), "Not Applicable",
     "Undefined"
     )
+| extend azone = properties.availabilityZone
```
implemented this change

wjpigott commented on build/template_kql/common/extend_resource.kql:
```diff
@@ -361,6 +384,8 @@ resources
     (type == 'microsoft.sql/servers/databases' and sku.tier == 'DataWarehouse'), "Not Applicable",
     (type == 'microsoft.dbformysql/servers'), tostring(properties.replicationRole),
     (type == 'microsoft.dbformysql/flexibleservers'), iif(properties.haEnabled == "Enabled", 'Not Applicable', tostring(properties.replicationRole)),
+    (type == 'microsoft.dbforpostgresql/servers'), tostring(properties.replicationRole),
+    (type == 'microsoft.dbforpostgresql/flexibleservers'), iif(properties.haEnabled == "Enabled", 'Not Applicable', tostring(properties.replicationRole)),
```
implemented this change

wjpigott commented on build/template_kql/common/extend_resource.kql:
```diff
@@ -384,6 +409,7 @@ resources
     Location = tolower(location),
     ResourceGroup = resourceGroup,
     Zones = tolower(tostring(zones)),
+    azone, 
```
implemented this change

wjpigott commented on build/template_kql/common/extend_resource.kql:
```diff
@@ -413,6 +439,7 @@ resources
     (Type == 'microsoft.sql/servers/databases'), "Not Applicable", 
     (Type == 'microsoft.documentdb/databaseaccounts'), iif(properties.backupPolicy.periodicModeProperties.backupStorageRedundancy == "Geo", "Healthy", "Unhealthy"),
     (Type == 'microsoft.dbformysql/servers'), iif(properties.storageProfile.geoRedundantBackup == "Enabled", "Healthy", "Unhealthy"),
+    (Type == 'microsoft.dbforpostgresql/servers'), iif(properties.storageProfile.geoRedundantBackup == "enabled", "healthy", "unhealthy"),
```
implemented this change

wjpigott commented on build/template_kql/common/extend_resource.kql:
```diff
@@ -421,4 +448,9 @@ resources
-    "Unknown")
\ No newline at end of file
+    "Unknown")
+| extend HighAvailability =  case(
+properties.highAvailability.mode == "ZoneRedundant", "Zone redundant",
+properties.highAvailability.mode == "SameZone", "Same zone",
+properties.highAvailability.mode)
+|extend GeoRedundantBackup = properties.backup.geoRedundantBackup  
```
implemented this change

wjpigott commented on build/template_kql/database/database_postgresqlflexible_resources_details.kql:
```diff
@@ -0,0 +1,3 @@
+${extend_resource}
+| project Type, SkuName, State, Location, ResourceGroup, Zones, azone, HighAvailability, GeoRedundantBackup, SubscriptionId, Name, ReplicationRole, Environment
+| where Type == 'microsoft.dbforpostgresql/flexibleservers'
```
Some of these columns do not exist in PostGresql I will match the output of mysqlflexible for the relevant columns

tsubasaxZZZ commented on build/template_kql/database/database_postgresqlflexible_resources_details_summary.kql:
In this logic, the value is negative if there is a Primary with more than one replication.
To calculate this value correctly, a JOIN is required, but JOINs are no longer available as the maximum number of JOINs used in the query so far is four.
Therefore, `NoReplica` column should be removed for PostgreSQL.

wjpigott replied:
Thank you!! This is removed.

tsubasaxZZZ commented on build/templates/databases.tpl.json:
PostgreSQL cannot distinguish between replication and non-replication, so do not use `Replication Role` with an Icon.
Only output the value of the Replication Role.

tsubasaxZZZ replied:
It doesn't seem that `Replication Role` icon is removed.
![image](https://github.com/Azure/reliability-workbook/assets/3822284/641896ce-3109-4e87-b618-e1c4e99d5291)

wjpigott replied:
Sorry that didnt get committed That is pushed now.

tsubasaxZZZ commented on build/template_kql/common/extend_resource.kql:
```diff
@@ -307,7 +329,21 @@ resources
+        properties.highAvailability.mode
+    ),
+    type == 'microsoft.dbforpostgresql/servers', "Not Applicable",
+    type == 'microsoft.dbforpostgresql/flexibleservers', case(
+        properties.highAvailability.mode == "Disabled" and properties.replicationRole != "Replica" , "Not Configured",
+        properties.replicationRole == "Replica", "Not Applicable",
+        properties.highAvailability.mode == "ZoneRedundant", "Zone Redundant",
+        properties.highAvailability.mode == "SameZone", "Same Zone",
+        properties.highAvailability.mode
+    ),
```
You need to update this part as follows because PostgreSQL replications role is different from MySQL:
```
    type == 'microsoft.dbforpostgresql/flexibleservers', case(
        properties.highAvailability.mode == "Disabled" and properties.replicationRole != "GeoAsyncReplica" , "Not Configured",
        properties.replicationRole == "GeoAsyncReplica", "Not Applicable",
        properties.highAvailability.mode == "ZoneRedundant", "Zone Redundant",
        properties.highAvailability.mode == "SameZone", "Same Zone",
        properties.highAvailability.mode
    ),
```

wjpigott replied:
Thank you this is updated.

tsubasaxZZZ commented on build/template_kql/common/extend_resource.kql:
```diff
@@ -413,6 +453,8 @@ resources
     (Type == 'microsoft.documentdb/databaseaccounts'), iif(properties.backupPolicy.periodicModeProperties.backupStorageRedundancy == "Geo", "Healthy", "Unhealthy"),
     (Type == 'microsoft.dbformysql/servers'), iif(properties.storageProfile.geoRedundantBackup == "Enabled", "Healthy", "Unhealthy"),
+    (Type == 'microsoft.dbforpostgresql/servers'), iif(properties.storageProfile.geoRedundantBackup == "enabled", "healthy", "unhealthy"),
+    (Type == 'microsoft.dbforpostgresql/flexibleservers'), properties.backup.geoRedundantBackup,
```
It's necessary that the resource is replicated or not, because PostgreSQL's replica cannot enable Backup.
![image](https://github.com/Azure/reliability-workbook/assets/3822284/161edd16-22a5-4e23-abfc-daca5ee2f432)

You can use the same code as like AutomaticFailover.

wjpigott replied:
Assuming you are meaning this code. 
`   type == 'microsoft.dbforpostgresql/flexibleservers', case(
        properties.highAvailability.mode == "Disabled" and properties.replicationRole != "GeoAsyncReplica" , "Not Configured",
        properties.replicationRole == "GeoAsyncReplica", "Not Applicable",
        properties.highAvailability.mode == "ZoneRedundant", "Zone Redundant",
        properties.highAvailability.mode == "SameZone", "Same Zone",
        properties.highAvailability.mode
    ),`

wjpigott commented on build/template_kql/common/extend_resource.kql:
```diff
@@ -307,7 +329,21 @@ resources
     (type == 'microsoft.documentdb/databaseaccounts'), "Not Applicable",
     type == 'microsoft.dbformysql/servers', "Not Applicable",
-    type == 'microsoft.dbformysql/flexibleservers', case(properties.haEnabled == "Enabled", "Configured", "Not Applicable"),
+     type == 'microsoft.dbformysql/flexibleservers', case(
+        properties.highAvailability.mode == "Disabled" and properties.replicationRole != "Replica" , "Not Configured",
+        properties.replicationRole == "Replica", "Not Applicable",
+        properties.highAvailability.mode == "ZoneRedundant", "Zone Redundant",
+        properties.highAvailability.mode == "SameZone", "Same Zone",
+        properties.highAvailability.mode
+    ),
```
reverted back to orginal mysql

wjpigott commented on build/template_kql/common/extend_resource.kql:
```diff
@@ -360,7 +398,9 @@ resources
     (type == 'microsoft.sql/servers/databases' and sku.tier == 'DataWarehouse'), "Not Applicable",
     (type == 'microsoft.dbformysql/servers'), tostring(properties.replicationRole),
-    (type == 'microsoft.dbformysql/flexibleservers'), iif(properties.haEnabled == "Enabled", 'Not Applicable', tostring(properties.replicationRole)),
+    (type == 'microsoft.dbformysql/flexibleservers'), tostring(properties.replicationRole),
```
reverted back to orginal mysql

wjpigott commented:
![image](https://github.com/Azure/reliability-workbook/assets/86477713/a4708f89-1ac2-4b48-82e8-7c1191b2c9fd)
All recommendations added except the one with the backup image as I am not sure what the goal is there. @tsubasaxZZZ

wjpigott commented:
@tsubasaxZZZ changes are complete for backup and other suggestions. Thank you.

wjpigott commented:
I don't use the editor as you are showing only after deploying it, so might have to learn how to do that :-)
So for small to Full I see that is changing from a 1 to a 3. 
    "size": 3,
    "title": "PostgreSQL - Flexible",

I think I have the icons set to unknown for the gray icon with (?) and as a 2 for the orange triangle.

From past conversations you said not to put any icons on Replication Role, so those are not set for me. 
Environment is still colors if you want that to be this way or not. 

I didnt have anything as unavailable to test with but set it properly I think.

![image](https://github.com/Azure/reliability-workbook/assets/86477713/75ef050d-f3ca-46b0-bfe6-e9ea0455b00e)

These changes are pushed up.

wjpigott commented:
@tsubasaxZZZ  also, how did you get the replication role to say geoasyncreplica vs.   "replicationRole": "AsyncReplica" as in the JSON output of the postgresql replica?

wjpigott commented:
On the GeoAsyncReplica I found a way to make this work by doing this in one line with !contains = "AsyncReplica" this seems to work. I had to change the sum on the GeoBackup to look for "not configured" vs. "disabled" as well.
![image](https://github.com/Azure/reliability-workbook/assets/86477713/3b37063b-c791-41a9-86f3-7b9a57c09654)
@tsubasaxZZZ

tsubasaxZZZ reviewed (changes requested):
It's necessary to revert to "disabled" for the UI to work properly and keep consistency.

wjpigott commented:
@tsubasaxZZZ  GeoBackup property has been changed to state disabled when the backup is not configured and not contain asyncreplica. Thanks! 
![image](https://github.com/Azure/reliability-workbook/assets/86477713/fa47cbe6-3385-4d79-bbf6-c1e2c9abd6a6)

tsubasaxZZZ reviewed (changes requested):
![image](https://github.com/Azure/reliability-workbook/assets/3822284/83e2300a-8f08-4637-9644-315a6967b8f4)

Summary and detail doesn't match because filtered by "not configured" on summary kql.

wjpigott commented:
Change has been pushed up for changing disabled for GeoBackup summary. 
![image](https://github.com/Azure/reliability-workbook/assets/86477713/0f60c295-74b2-4c3d-9868-4ceb10501074)
@tsubasaxZZZ

tsubasaxZZZ commented:
@wjpigott Could you revert to `build/template_kql/storage/storage_account_resources_details_summary.kql`?
You need to consider that Storage Accounts are deployed whether or not they are in the 3+0 region.

You don't need to modify `build/template_kql/storage/storage_account_resources_details_summary.kql`. And you need to fix condition of the part: https://github.com/Azure/reliability-workbook/blob/e4ee6edc13e0d709cfed73feb9970dc8ccc4a3d9/build/template_kql/common/extend_resource.kql#L311C1-L315C11
Please add 3+0 region to `in` statement. You can check the target region the following document:
https://learn.microsoft.com/en-us/azure/reliability/cross-region-replication-azure#regions-with-availability-zones-and-no-region-pair

And please create new PR to fix this first to identify other issue.

wjpigott commented:
> @wjpigott Could you revert to `build/template_kql/storage/storage_account_resources_details_summary.kql`? You need to consider that Storage Accounts are deployed whether or not they are in the 3+0 region.
> 
> You don't need to modify `build/template_kql/storage/storage_account_resources_details_summary.kql`. And you need to fix condition of the part: https://github.com/Azure/reliability-workbook/blob/e4ee6edc13e0d709cfed73feb9970dc8ccc4a3d9/build/template_kql/common/extend_resource.kql#L311C1-L315C11 Please add 3+0 region to `in` statement. You can check the target region the following document: https://learn.microsoft.com/en-us/azure/reliability/cross-region-replication-azure#regions-with-availability-zones-and-no-region-pair
> 
> And please create new PR to fix this first to identify other issue.

@tsubasaxZZZ  This was reverted back and pushed up. I will let the storage owner make the other changes to ZRS to make this acceptable when time permits. Thanks!
